unit/analyses/call_graph.cpp

```diff
@@ -0,0 +1,113 @@
+
+#include <analyses/call_graph.h>
+
+#include <util/symbol_table.h>
+#include <util/std_code.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+SCENARIO("call_graph",
+  "[core][util][call_graph]")
+{
+
+  symbol_tablet symbol_table;
+
+  code_typet void_function_type;
+
+  {
+    code_blockt calls;
+    code_function_callt call1;
+    call1.function()=symbol_exprt("A", void_function_type);
+    code_function_callt call2;
+    call2.function()=symbol_exprt("B", void_function_type);
+    calls.move_to_operands(call1);
+    calls.move_to_operands(call2);
+
+    symbolt a_function;
+    a_function.name="A";
+    a_function.type=void_function_type;
+    a_function.mode=ID_java;
+    a_function.value=calls;
+    symbol_table.add(a_function);
+  }
+
+  {
+    code_blockt calls;
+    code_function_callt call1;
+    call1.function()=symbol_exprt("C", void_function_type);
+    code_function_callt call2;
+    call2.function()=symbol_exprt("D", void_function_type);
+    calls.move_to_operands(call1);
+    calls.move_to_operands(call2);
+
+    symbolt b_function;
+    b_function.name="B";
+    b_function.type=void_function_type;
+    b_function.mode=ID_java;
+    b_function.value=calls;
+    symbol_table.add(b_function);
+  }
+
+  {
+    symbolt c_function;
+    c_function.name="C";
+    c_function.type=void_function_type;
+    c_function.mode=ID_java;
+    c_function.value=code_skipt();
+    symbol_table.add(c_function);
+  }
+
+  {
+    symbolt d_function;
+    d_function.name="D";
+    d_function.type=void_function_type;
+    d_function.mode=ID_java;
+    d_function.value=code_skipt();
+    symbol_table.add(d_function);
+  }
+
+  call_grapht call_graph_from_symbol_table(symbol_table);
+  call_grapht inverse_call_graph_from_symbol_table=
+    call_graph_from_symbol_table.get_inverted();
+
+
+  GIVEN("A call graph constructed from a symbol table")
+  {
+    WHEN("Checking the call graph")
+    {
+      REQUIRE(call_graph_from_symbol_table.graph.size()==4);
+      REQUIRE(call_graph_from_symbol_table.graph.count("A")==2);
+      REQUIRE(call_graph_from_symbol_table.graph.count("B")==2);
+    }
+
+    WHEN("Checking the inverted call graph")
+    {
+      REQUIRE(inverse_call_graph_from_symbol_table.graph.size()==4);
+      REQUIRE(inverse_call_graph_from_symbol_table.graph.count("A")==1);
+      REQUIRE(inverse_call_graph_from_symbol_table.graph.count("B")==1);
+      REQUIRE(inverse_call_graph_from_symbol_table.graph.count("C")==1);
+      REQUIRE(inverse_call_graph_from_symbol_table.graph.count("D")==1);
+    }
+  }
+
+  goto_functionst goto_functions;
+  stream_message_handlert msg(std::cout);
+  goto_convert(symbol_table, goto_functions, msg);
+
+  call_grapht call_graph_from_goto_functions(goto_functions);
+  call_grapht inverse_call_graph_from_goto_functions=
+    call_graph_from_goto_functions.get_inverted();
+
+  GIVEN("A call graph constructed from GOTO functions")
+  {
+    REQUIRE(call_graph_from_symbol_table.graph==
```

Comment:
I'd be a bit nervous about using the equality on a complex type and would slightly prefer copy/pasting the REQUIRES from above, but if your confident in it then don't worry.